Move system smtp to secret

miguelsorianod opened the pull request:
This PR changes the deployment of the SMTP configuration from a ConfigMap to a Secret.
The reason of this is because some of the configuration contains credentials.

miguelsorianod commented:
Pending to implement the upgrade procedure but the current PR state should have the needed code in templates and operator to deploy SMTP configuration with a secret instead of a configmap.
@eguzki you can review this part for the moment

miguelsorianod commented:
Updated PR to include upgrade process to migrate from the 'smtp' ConfigMap to the 'system-smtp' Secret.
In the future the order/coordination in how the upgrade is performed might have to be changed to have consistent changes

eguzki commented:
looking good

miguelsorianod commented:
Currently the upgrade procedure should work because there isn't an upgrade step at this moment that requires "atomicity" of the changes.
How do you want to proceed? We can merge this and work on a solution when the need arises/handle it in a separate PR or specify what we should solve in this PR before merging it.

eguzki commented:
missing smtp secret documentation

miguelsorianod commented:
Added system-smtp secret documentation.
Ready for re-review @eguzki

miguelsorianod commented:
I've noticed the APIManager controller unit test related to the upgrade fails. The reason for that is that the upgrade code expects the "system" ConfigMap to exist (see migrateSystemSMTPData method in upgrade.go) but it does not exist anymore due to it has been removed in this PR being replaced by a secret.

Should we make the UpgradeApiManager implement an interface and use an interface instead from the apimanager_controller being the upgrader a field of the ReconcileAPIManager struct? and mock it in the test so the upgrade method returns an arbitrary result? What do you think?

eguzki commented:
Why not just create the smtp configmap for the upgrade test? you can even test the config map has been replaced by secret.

miguelsorianod commented:
> Why not just create the smtp configmap for the upgrade test? you can even test the config map has been replaced by secret.

I thought about that but this means that in the next release the test will break/need to be changed so they would be prone to errors. Also the number of calls to the Reconcile method in the test will vary depending on the number of elements to upgrade.

In any case I've implemented it so you can see how the addition of this new functionality has affected the test in the apimanager_controller_test (in a new commit). I've also modified the upgrade.go implementation when creating the system-smtp secret because it was being created without the controllerreference and now it sets the secret data using Data instead of StringData field due to when mocking the API calls the StringData seems that is not automatically converted to Data when creating the object and this caused tests to fail too. You can re-review that file too.

eguzki commented:
In the next release, all the upgrade tests do not make sense at all and have to be "removed", as it has to be done with the upgrade code.

eguzki commented on doc/apimanager-reference.md:
```diff
@@ -326,3 +326,15 @@ existing secret names.
 | AWS_ACCESS_KEY_ID | AWS Access Key ID to use in S3 Storage for System's file storage | N/A |
 | AWS_SECRET_ACCESS_KEY | AWS Access Key Secret to use in S3 Storage for System's file storage | N/A |
+
+#### system-smtp
```
Recently some user complained about the lack of documentation about SMTP configuration. Do you think it is enough to document a table with available fields in the apimanager reference?

thomasmaas replied:
What are the alternatives?

thomasmaas commented on doc/apimanager-reference.md:
```diff
@@ -326,3 +326,15 @@ existing secret names.
+| **Field** | **Description** | **Default value** |
+| --- | --- | --- |
+| address | Address (hostname or IP) of the remote mail server to use. If set to a value different than `""` System will use the mail server to send mails related to events that happen in the API management solution |  `""` |
+| username | In case the mail server requires authentication, set the username using this setting | `""` |
```
```suggestion
| username | In case the mail server requires authentication | `""` |
```

miguelsorianod replied:
:+1:

thomasmaas commented on doc/apimanager-reference.md:
```diff
@@ -326,3 +326,15 @@ existing secret names.
+| --- | --- | --- |
+| address | Address (hostname or IP) of the remote mail server to use. If set to a value different than `""` System will use the mail server to send mails related to events that happen in the API management solution |  `""` |
+| username | In case the mail server requires authentication, set the username using this setting | `""` |
+| password | In case the mail server requires authentication, set the password using this setting | `""` |
```
```suggestion
| password | In case the mail server requires authentication. | `""` |
```

miguelsorianod replied:
:+1:

thomasmaas commented on doc/apimanager-reference.md:
```diff
@@ -326,3 +326,15 @@ existing secret names.
+| address | Address (hostname or IP) of the remote mail server to use. If set to a value different than `""` System will use the mail server to send mails related to events that happen in the API management solution |  `""` |
+| username | In case the mail server requires authentication, set the username using this setting | `""` |
+| password | In case the mail server requires authentication, set the password using this setting | `""` |
+| domain | If a HELO domain is needed to be specified set it using this setting | `""` |
```
```suggestion
| domain | In case the mail server requires a HELO domain. | `""` |
```

miguelsorianod replied:
:+1:

thomasmaas commented on doc/apimanager-reference.md:
```diff
@@ -326,3 +326,15 @@ existing secret names.
+| password | In case the mail server requires authentication, set the password using this setting | `""` |
+| domain | If a HELO domain is needed to be specified set it using this setting | `""` |
+| port | Port of the remote mail server to use | `""` |
+| authentication | In case the mail server requires authentication, set this setting to the authentication type here. `plain` to send the password in the clear, `login` to send password Base64 encoded, or `cram_md5` to combine a Challenge/Response mechanism to exchange information and a cryptographic Message Digest 5 algorithm to hash important information | `""` |
```
```suggestion
| authentication | Specify the authentication type in case the mail server requires authentication. Use `plain` to send the password in the clear, `login` to send password Base64 encoded, or `cram_md5` to combine a Challenge/Response mechanism to exchange information and a cryptographic Message Digest 5 algorithm to hash important information | `""` |
```

thomasmaas replied:
But "or `cram_md5` to combine a Challenge/Response mechanism to exchange information and a cryptographic Message Digest 5 algorithm to hash important information" is very hard to read, can we simplify?

thomasmaas replied:
would you use bother username and password fields when picking `cram_md5`?

miguelsorianod replied:
That description is the one from the official rubydocs documentation. Anyway, I've changed it to '`cram_md5` to combine a Challenge/Response mechanism based on the HMAC-MD5 algorithm '

If you don't need to set username and password then you don't have to in that case.
I've updated the username and password fields to
"In case the mail server requires authentication and the authentication type requires it"
instead of just
"In case the mail server requires authentication"

thomasmaas commented on doc/apimanager-reference.md:
```diff
@@ -326,3 +326,15 @@ existing secret names.
+| domain | If a HELO domain is needed to be specified set it using this setting | `""` |
+| port | Port of the remote mail server to use | `""` |
+| authentication | In case the mail server requires authentication, set this setting to the authentication type here. `plain` to send the password in the clear, `login` to send password Base64 encoded, or `cram_md5` to combine a Challenge/Response mechanism to exchange information and a cryptographic Message Digest 5 algorithm to hash important information | `""` |
+| openssl.verify.mode | When using TLS, you can set how OpenSSL checks the certificate. This is really useful if you need to validate a self-signed and/or a wildcard certificate. You can use the name of an OpenSSL verify constant: `none` or `peer` | `""` |
```
```suggestion
| openssl.verify.mode | When using TLS, use this field to specify how OpenSSL checks the certificate. This is useful to validate a self-signed and/or a wildcard certificate. Use the name of an OpenSSL verify constant: `none` or `peer` | `""` |
```

thomasmaas replied:
<img width="181" alt="Screenshot 2019-12-18 18 42 19" src="https://user-images.githubusercontent.com/54224/71109738-29fe6300-21c6-11ea-8b77-13ece463bac2.png">

wouldn't it make sense to re-order and rename the fields to:
- address
- port
- helo.domain
- auth.type
- auth.username
- auth.password
- openssl.verify.mode

miguelsorianod replied:
Ok to the order.
You cannot rename the fields as they are the names that System looks.

miguelsorianod commented:
system-smtp documentation table updated
